회원가입 구현

jaea-kim opened the pull request:
## 구현 내용
**회원가입(이름, 아이디, 이메일, 비밀번호) API**
- 이메일 인증을 완료해야만 회원가입이 가능하도록 구현
- 아이디/이메일 중복 체크 API 제공

**이메일 인증**
- 인증 요청 시, 토큰 생성 후 메일로 인증 링크 발송
- 인증 링크 클릭 시 DB에 인증 완료 처리
- 회원가입이 완료되면 DB에 토큰 삭제
- 구글 SMTP, JavaMailSender 사용
![스크린샷 2025-06-22 오후 8 58 09](https://github.com/user-attachments/assets/2128d6ae-d7fc-4217-adcc-fe16f947aac3)


## 추후에
- 비밀번호 암호화
- 이메일 토큰 만료 기간 설정 및 재전송

## 고민
이메일 인증 - 회원 DB 저장 방식 : 인증되지 않은 유저도 유저DB에 저장해 유저 DB에 인증 컬럼을 만들고 인증 여부를 표시해야하나 고민했습니다.. 
제 생각에는 이메일 인증을 위해서는 발급된 인증 토큰을 저장할 공간이 필요했는데 그러기 위해서는 토큰 테이블을 생성하고 관리해야 했습니다. 그러면 굳이 유저 DB에 인증여부를 표시하는 것이 중복되어 저장한다고 생각이 들어서 이메일 인증이 되지 않으면 회원가입을 하지 못하도록해 유저DB에는 인증이 완료된 유저만 저장되도록 했습니다. 
이메일 인증 테이블을 생성해서 해당 테이블에 인증 여부와 토큰를 저장하였습니다. 회원가입이 완료되면 해당 테이블에 해당 이메일을 삭제하도록 했습니다. 
이렇게 구현하는 부분에 있어 문제가 없을까요? 

### 테스트 관련
1. 현재 서비스만 테스트 코드를 작성했는데 컨트롤러와 레포지토리에서 딱히 로직이 없는데 어떤 걸 테스트해야하는지 모르겠습니다..ㅠㅠ
2. 현재 작성한 단위 테스트 코드에서 테스트를 빼먹은 부분은 없는지 계속 고민이 되는데 테스트 커버리지로 판단을 하는건가요..?

jaea-kim committed:
feat: 아이디 중복 검사 #5

jaea-kim committed:
feat: 일반 회원가입 구현 #5

- 컨트롤러, 서비스, 요청DTO 구현
- ErrorResponse에 직렬화를 위한 Getter 추가

jaea-kim committed:
feat: User 엔티티 작성 #5

jaea-kim committed:
feat: 비밀번호 암호화 #5

jaea-kim committed:
feat: 중복 에러코드 추가 및 파라미터 예외 추가 #5

jaea-kim committed:
feat: 이메일 전송 설정 추가 #5

jaea-kim committed:
feat: 이메일 전송, 검증 예외 추가 #5

jaea-kim committed:
feat: 이메일 인증 요청 추가 #5

- post 요청 시 이메일 인증 토큰 발급 및 이메일 발송
- get 요청 시 토큰 검증

jaea-kim committed:
feat: 회원가입 시 이메일 인증 확인 추가 및 완료 후 이메일 토큰 삭제 #5

jaea-kim committed:
feat: 비회원 접근이 필요한 API 설정 #5

jaea-kim committed:
refactor: 테스트 패키지 경로 수정 및 로깅 레벨 수정 #3

jaea-kim committed:
test: 서비스 레이어 테스트 작성 #5

jaea-kim committed:
chore: println 삭제 #5

jaea-kim committed:
refactor: 이메일 인증 URL 변경 #5

derek-Flab commented on src/main/java/com/project/triplog/controller/UserController.java:
```diff
@@ -0,0 +1,43 @@
+    private final EmailVerificationService emailVerificationService;
+
+    @GetMapping("/users/validation")
+    public ResponseEntity<Boolean> checkUsername(@RequestParam String username) {
```
1. `Boolean`으로 응답객체가 만들기 보다 
응답 JSON객체가 내려가는게 좋을 것 같습니다. result라는 네이밍은 단순한 예시입니다. 
ASIS: true
TOBE: 
{
  "result": false
}

---

derek-Flab replied:
지금 이 validation이라는 네이밍이 좀 모호한 것 같습니다. 
지금 이 서버에 이름이 있는지 조회하는거잖아요? 그러면 exist라던가 이런게 더 맞을것 같습니다.

근데 좀 더 생각해보면 지금 사용자 회원가입 입력값을 검증하는 것이라면?

1. checkUsername 네이밍 이건 범용적인 이름으로 당연히 바꾸어야 할 것 같고요. 
2. 요청값도 request body로 받도록 하는게 맞을 것 같습니다. 사용자의 중요정보가 get파라미터에 노출되지 않도록 하는것도 필요할 것 같습니다.

jaea-kim replied:
validation이라는 네이밍은 수정하겠습니다! 

1. 사용자 회원가입 입력값 중에 username과 email을 각각 검증하게 되는데 같은 컨트롤러에서 받을 수 있도록 하는건가요?
범용적인 이름이라고 하는 것이 username만 검증하더라도 크게 보면 사용자 회원가입 입력값을 검증하는거라 그런 의미에서 범용적인 이름으로 변경하는 게 좋다는 걸까요?  
2. GET 요청시에는 request body포함되지 않아야 된다고 알고 있어서 저렇게 진행했는데 최대한 body를 사용하지 않도록 하도록 하되 body로 받아도 된다? 이거 일까요..? 아님 GET 요청으로 받는 것이 적절하지 않은걸까요?

derek-Flab replied:
1. 아 이미 지어주신 이름이 약간 범용적인거 같아서요! validation에 request parameter가 계속 추가하실건 염두해두신거죠? 아닌가요? 따로따로 해도 상관은 없지만, 여러 값을 검증할때마다 API를 새로만드는게 괜찮을정도로 1-2개밖에 검증할게 없다면 괜찮을 것 같습니다.
2. 검증로직은 GET요청이 아닌 POST로 선언하고 관리하는게 맞지 않을까라는 의미였습니다!

derek-Flab commented on src/main/java/com/project/triplog/controller/UserController.java:
```diff
@@ -0,0 +1,43 @@
+        return ResponseEntity.ok().build();
+    }
+
+    @PostMapping("/users/email/verification")
```
이거 email 인증 요청을 하는거죠? 

`@PostMapping("/users/email/verification")`
`@GetMapping("/users/email/verification")`

위의 꺼는 검증을 요청하고
아래꺼는 검증값을 검증한다.

뭐 이런거 같은데 조금 모호한것같습니다.

검증을 요청한다는 `request-verification`이 더 낫지 않을가요? 
아래는 검증라는 것에는 잘 어울리나, 검증은 제한 시간 같은 것도 있을텐데 get메서드도 모호해보입니다. 
POST가 어떨까요

derek-Flab replied:
이것들도 마찬가지로 응답이 명확히 객체여야 합니다.

jaea-kim replied:
Post 요청은 email 인증 요청을 보내고 
get 요청은 email 인증 요청에서 받은 token을 검증한다. 이렇게 진행했습니다. 

get 메서드와 url은 수정하겠습니다. 

요청에는 항상 응답이 객체로 와야하나요? 
검증이 완료됐다는 상태코드를 보냈기에 응답이 없어도 된다고 생각해서 그렇게 진행했는데 응답이 객체로 와야한다면 메시지를 포함한 응답 객체가 가야하는걸까요? 
항상 객체가 응답이 가야한다면 예외 형식을 지정한 것처럼 응답 형식도 지정해서 사용하는 것이 좋을까요?

derek-Flab commented on src/main/java/com/project/triplog/service/EmailVerificationService.java:
```diff
@@ -0,0 +1,81 @@
+
+	public void sendVerificationEmail(EmailRequest emailRequest) {
+		String token = generateToken(emailRequest.getEmail());
+		String verificationUrl = "http://localhost:8080/users/email/verification?email="
```
http://localhost:8080
요 host는 환경별로 달라질 수 있기 때문에 환경변수 처리가 필요해보입니다.

derek-Flab replied:
이메일 인증에 시간 제한 같은것은 없나요?

jaea-kim replied:
네! 해당 부분은 수정하겠습니다. 

이메일 인증에 시간 제한을 두고 싶었는데 거기까지는 진행을 못했습니다. 
일단, 전체적으로 진행을 한 뒤에 추가로 시간 제한을 구현하려고 했습니다..

derek-Flab commented on src/main/java/com/project/triplog/service/UserService.java:
```diff
@@ -0,0 +1,47 @@
+	}
+
+	private String encodingPassword(String password) {
+		return passwordEncoder.encode(password);
```
UserService에 패스워드 인코딩? 자신의 책임이 아닌거 같습니다.

jaea-kim replied:
패스워드 인코딩이 컨트롤러와 도메인의 책임이 아니라고 생각했습니다. 
그래서 서비스에서 인코딩을 해서 도메인에서 넘겨줘야한다고 생각해서 진행했는데.. 

인코딩은 보안?에 관한 사항이라 회원가입과 로그인 등을 처리하는 서비스의 책임과는 다르다고 봐야하는걸까요..? 

서비스와 유저 도메인 사이에 유저를 생성하는 객체를 만들고 거기서 인코딩을 담당하게 하는게 좋을까요?

derek-Flab replied:
이전에 드린 리뷰에서 제가 코드를 조금 과하게 해석했던 것 같습니다.
다시 보니 passwordEncoder.encode(password) 부분은 별도의 메서드로 분리하지 않아도 충분해 보입니다.
이렇게 분리하다 보니 UserService가 마치 패스워드 암호화 책임까지 맡고 있는 것처럼 보일 수 있었던 것 같습니다.
실제로는 단순히 메서드 분리가 과했던 부분으로 이해하시면 될 것 같습니다.

derek-Flab commented on src/main/java/com/project/triplog/service/UserService.java:
```diff
@@ -0,0 +1,47 @@
+	public void join(JoinRequest joinRequest) {
+		emailVerificationService.checkVerified(joinRequest.getEmail());
+		if (isExistUsername(joinRequest.getUsername())) {
+			throw new DuplicationException("이미 존재하는 사용자 이름입니다.");
```
사용자 이름은 얼마든지 중복될 수 있는데, 이미 사용자 이름이 있으면 가입이 안되는건가요?

jaea-kim replied:
아 저는 username 을 아이디 처럼 생각해서 중복되지 않게 해야한다고 생각했습니다. 

username 과 name을 따로 두고 username 을 id라고 생각하고 name을 이름이라고 생각했는데 이미 이메일도 중복불가능하게 사용하는데 username까지 사용하는 건 너무 불필요할까요?
현재는 이메일은 휴대폰 인증처럼 인증하는 용도로 생각했고 username을 id라고 생각해 username과 비밀번호를 입력해서 로그인하도록 구현했습니다.

derek-Flab replied:
username이 id? 라면 뭔가 네이밍이 헷갈리게 되버릴꺼 같네요! 그렇다면 userId가 맞지 않을까요?
username은 일반적으로 "홍길동"이런 이름처럼 충분히 중복이 가능한 필드라고 여겨집니다!

jaea-kim committed:
feat: 사용자 중복 확인 API 개선 #5

- username/email 중복 확인을 GET에서 POST로 변경
- UsernameCheckRequest, EmailCheckRequest DTO 추가

jaea-kim committed:
feat: 이메일 인증 API 개선 #5

- URL 패턴을 명확하게 분리
- GET에서 POST로 변경

jaea-kim committed:
refactor: username 필드를 userId로 변경 #5

- User 도메인 클래스의 username 필드를 userId로 변경
